log.go

```diff
@@ -190,3 +190,24 @@ func emitLogStoreMetrics(s LogStore, prefix []string, interval time.Duration, st
 		}
 	}
 }
+
+type CommitTrackingLogStore interface {
+	LogStore
+
+	// StageCommitIndex stages a new commit index to be persisted.
+	// The staged commit index MUST only be persisted in a manner that is atomic
+	// with the following StoreLogs call in the face of a crash.
+	// This allows the Raft implementation to optimize commit index updates
+	// without risking inconsistency between the commit index and the log entries.
+	//
+	// The implementation MUST NOT persist this value separately from the log entries.
+	// Instead, it should stage the value to be written atomically with the next
+	// StoreLogs call.
+	//
+	// GetCommitIndex MUST never return a value higher than the last index in the log,
+	// even if a higher value has been staged with this method.
+	//
+	// idx is the new commit index to stage.
+	StageCommitIndex(idx uint64) error
+	GetCommitIndex() (uint64, error)
```

Comment:
How would a `GetCommitIndex()` could be implemented in a real store? Would it read the latest stored log and return the commit index associated to it? What if it don't find any because those logs were stored using a store that don't support fast-recovery?

Reply:
For BoltDB, I imagine commit index would be a single KV in a separate bucket from logs so it would just read that and return it.

For WAL I anticipated extending the format slightly so that each commit entry in the log stores the most recently staged commit index and then re-populated that into memory when we open the log an scan it like we do with indexes.

If there is no commit index stored, we should just return `0, nil` which is always safe and has the same behavior as current code I think right?

Reply:
> If there is no commit index stored, we should just return 0, nil which is always safe and has the same behavior as current code I think right?

I agree! I think that should be documented though. Because the API allow erroring on `GetCommitIndex()` it could easily be mistaken as a possible error case.